### ISSUE TITLE
Cython backend doesn't behave the same as cffi/c

### ISSUE THREAD

**Issue by pvanderlinden** [zeromq/pyzmq#652]
It is impossible to send a zero length string with send when using the cython backend (To close a connection with a STREAM socket), when doing the same with cffi it works fine.

example:
```
  File "server.py", line 33, in got_request
    socket.send_multipart([identity, b''])
  File "/lib/python2.7/site-packages/zmq/sugar/socket.py", line 278, in send_multipart
    return self.send(msg_parts[-1], flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 574, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:5434)
  File "zmq/backend/cython/socket.pyx", line 621, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:5196)
  File "zmq/backend/cython/socket.pyx", line 182, in zmq.backend.cython.socket._send_copy (zmq/backend/cython/socket.c:2044)
  File "zmq/backend/cython/checkrc.pxd", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:6248)
zmq.error.ZMQError: Bad address
```

**Comment by minrk** [comment]
What pyzmq, libzmq versions? Can you provide a complete example that reproduces the problem and shows a difference in behavior between Cython and CFFI?

**Comment by pvanderlinden** [comment]
pyzmq version: 14.5.0
zmq version: 4.0.4

If you run the following script (taken from the example how to use the ZMQ_STREAM socket):
```python
import zmq

context = zmq.Context()
socket = context.socket(zmq.STREAM)
socket.bind('tcp://127.0.0.1:8000')

while True:
    identity = socket.recv()
    req = socket.recv()
    print req
    socket.send(identity, zmq.SNDMORE)
    socket.send('HTTP/1.1 200 OK\r\nConnection: Close\r\nContent-length: 11\r\n\r\nHello World')
    # Close Connection
    socket.send(identity, zmq.SNDMORE)
    socket.send('')
```

With cffi it works fine: it returns a http response and closes the connection, then accepts the next request. With the cython backend you get the Exception above.

It is probably also better to default to cffi, as it seems to be the move in the python community to use for c extensions.

**Comment by minrk** [comment]
> It is probably also better to default to cffi, as it seems to be the move in the python community to use for c extensions.

The Cython backend is generally more reliable and more efficient and preferable to CFFI, except on PyPy, so I don't think the default should switch on CPython.

**Comment by minrk** [comment]
So far, I've found a bug in the CFFI backend, which is failing to check the returncode that Cython is checking. The CFFI backend is actually seeing the same Bad Address error, it's just not reporting it. Since it doesn't affect any other socket type, I think this is going to prove to be a bug in the underlying STREAM socket in libzmq.

**Comment by pvanderlinden** [comment]
Should I open this bug at libzmq for it?

**Comment by minrk** [comment]
@pvanderlinden looks like the bug's already been fixed in libzmq, and there's already a test for it. Apparently [I even wrote it](https://github.com/zeromq/libzmq/pull/721) and completely forgot. From the comments, it looks like the fix was supposed to be backported to 4.0, but never was.

**Comment by minrk** [comment]
empty frame fix backported to zeromq 4.0.x in zeromq/zeromq4-x#107. If you use the 4.1rc, it should have no problem.
